tools/sort.py

```diff
@@ -6,127 +7,586 @@
+        print ("Sorting by face similarity...")
         
-        img_list = [ [x, cv2.calcHist([cv2.imread(x)], [0], None, [256], [0, 256]) ] for x in tqdm( self.find_images(input_dir), desc="Loading") ]
+        img_list = [ [x, face_recognition.face_encodings(cv2.imread(x)) ] for x in tqdm( self.find_images(input_dir), desc="Loading") ]
 
         img_list_len = len(img_list)
         for i in tqdm ( range(0, img_list_len-1), desc="Sorting"):
             min_score = 9999.9
             j_min_score = i+1
             for j in range(i+1,len(img_list)):
-                score = cv2.compareHist(img_list[i][1], img_list[j][1], cv2.HISTCMP_BHATTACHARYYA)
+            
+                f1encs = img_list[i][1]
+                f2encs = img_list[j][1]
+                if f1encs is not None and f2encs is not None and len(f1encs) > 0 and len(f2encs) > 0:
+                    score = face_recognition.face_distance(f1encs[0], f2encs)[0]
+                else: 
+                    score = 9999.9
+                
                 if score < min_score:
                     min_score = score
                     j_min_score = j            
             img_list[i+1], img_list[j_min_score] = img_list[j_min_score], img_list[i+1]
             
-        self.process_final_rename (input_dir, img_list)
+        self.process_final_rename (output_dir, img_list)
                 
         print ("Done.")
-        
-    def process_face(self):
+
+    def process_face_cnn(self):
+        from lib import FaceLandmarksExtractor
+
         input_dir = self.arguments.input_dir
-        
+        output_dir = self.arguments.output_dir
+
         print ("Sorting by face similarity...")
-        
+
+        img_list = []
+        for x in tqdm( self.find_images(input_dir), desc="Loading"):
+            d = FaceLandmarksExtractor.extract(cv2.imread(x), 'cnn', True)
+            img_list.append( [x, np.array(d[0][1]) if len(d) > 0 else np.zeros ( (68,2) ) ] )
+
+        img_list_len = len(img_list)
+        for i in tqdm ( range(0, img_list_len-1), desc="Sorting"):
+            min_score = 9999999
+            j_min_score = i+1
+            for j in range(i+1,len(img_list)):
+
+                fl1 = img_list[i][1]
+                fl2 = img_list[j][1]
+                score = np.sum ( np.absolute ( (fl2 - fl1).flatten() ) )
+
+                if score < min_score:
+                    min_score = score
+                    j_min_score = j
+            img_list[i+1], img_list[j_min_score] = img_list[j_min_score], img_list[i+1]
+
+        self.process_final_rename (output_dir, img_list)
+
+        print ("Done.")
+
+    def process_face_dissim(self):
+        input_dir = self.arguments.input_dir
+        output_dir = self.arguments.output_dir
+
+        print ("Sorting by face dissimilarity...")
+
         img_list = [ [x, face_recognition.face_encodings(cv2.imread(x)) ] for x in tqdm( self.find_images(input_dir), desc="Loading") ]
```

Comment:
`img_list = [ [x, face_recognition.face_encodings(cv2.imread(x)), 0 ] for x in tqdm( self.find_images(input_dir), desc="Loading") ]`